runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -359,78 +359,40 @@ end
 ---@param bufnr number
 ---@param client_id number|nil If nil, then return all of the diagnostics.
 ---                            Else, return just the diagnostics associated with the client_id.
-function M.get(bufnr, client_id)
-  if client_id == nil then
-    local all_diagnostics = {}
-    for iter_client_id, _ in pairs(diagnostic_cache[bufnr]) do
-      local iter_diagnostics = M.get(bufnr, iter_client_id)
-
-      for _, diagnostic in ipairs(iter_diagnostics) do
-        table.insert(all_diagnostics, diagnostic)
-      end
-    end
-
-    return all_diagnostics
-  end
-
-  return diagnostic_cache[bufnr][client_id] or {}
-end
-
---- Get the diagnostics by position
----
----@param bufnr number|nil The buffer number
----@param position table|nil The (0,0)-indexed position
 ---@param opts table|nil Configuration keys
 ---         - severity: (DiagnosticSeverity, default nil)
 ---             - Only return diagnostics with this severity. Overrides severity_limit
 ---         - severity_limit: (DiagnosticSeverity, default nil)
 ---             - Limit severity of diagnostics found. E.g. "Warning" means { "Error", "Warning" } will be valid.
----@param client_id number|nil the client id
----@return table Table with map of line number to list of diagnostics.
---               Structured: { [1] = {...}, [5] = {.... } }
-function M.get_position_diagnostics(bufnr, position, opts, client_id)
+---@param predicate function|nil Optional function for filtering diagnostics
+---@param comp      function|nil Optional function for sorting diagnostics
+function M.get(bufnr, client_id, opts, predicate, comp)
```

Comment:
I'd leave out the `opts` parameter. The predicate can be used to filter based on the severity - there is no need to have a special case it.

I'd also remove the `comp` parameter - given that it is at the end, the call-site can do the `table.sort` call on the result - no need to inject the logic into `M.get`.

Reply:
Another alternative might be to just keep `M.get()` as is, and do a `vim.tbl_filter(predicate, M.get(..))` at the call-site.
In some cases it may require an additional copy to create the filtered table, but not sure if that is a concern.

@mjlbach wdyt?

Reply:
I like your suggestion better @mfussenegger because it simplifies the interfacet to M.get but I don't know offhand if the perf characteristics would be worse given this is a large copy

Reply:
@mfussenegger I've removed opts and comp from the function signature but left `predicate` as a parameter :slightly_smiling_face: - lmk if this works. I can delegate the predicate logic to the call-site as well if you prefer